### PR TITLE
Global Grid Variable Dimension

### PR DESCRIPTION
**Pull request by sslattery**
Adds the general dimension template features to `GlobalGrid` with a 2D-specific unit test. Note that there are no `Edge` entity types in 2D - only `Cell`, `Face`, and `Node`. In the implementation `enable_if` is used to enable/disable function implementations that would be specific to a given dimension.

Part of #340. Depends on #338

### CODE REVIEW

**Comment by sslattery** [comment]
This is ready for review but I will leave as WIP until the upstream PRs are merged.

**Comment by streeve** [comment]
In 2D I would expect to have an `Edge` and no `Face` - am I missing something?

**Comment by sslattery** [comment]
> In 2D I would expect to have an `Edge` and no `Face` - am I missing something?

Good question! One canonical way to do this is have `Cell` be the entity with topological dimension equal to space dimension and `Face` be the entity with topological dimension equal to space dimension minus 1. So in 3D, Cell is dim 3, face dim 2, edge dim 1, and node dim 0. In 2D, Cell is dim 2, face dim 1, and node dim 0.

This isn't the only way to do this but I have seen this structure in many mesh libraries and it also fits well with the nomenclature used in a lot of discretization.

Documentation should be improved to reflect this.

**Comment by sslattery** [comment]
@streeve This PR is next in line for #340

**Comment by sslattery** [comment]
retest this please

**Review by streeve — APPROVED** [review]
Maybe some comments saying these tests shouldn't be consolidated to test the overloads like you've mentioned. And if keeping face vs edge is common I'm okay with it

File: cajita/src/Cajita_GlobalGrid.hpp [hunk 1]
```diff
@@ -76,17 +76,33 @@ class GlobalGrid
     // Get the MPI rank of a block with the given indices. If the rank is out
     // of bounds and the boundary is not periodic, return -1 to indicate an
     // invalid rank.
-    int blockRank( const int i, const int j, const int k ) const;
+    int blockRank( const std::array<int, num_space_dim>& ijk ) const;
```
[the remark this answers is not in the document]
Comment (sslattery) [reply cajita/src/Cajita_GlobalGrid.hpp:79]: I think we should leave them. They create a nice interface that is dimension-specific with a thin wrapper but also allow for potentially dimension-agnostic code to be written.

**Pull request merged**